File: js/src/ebook.mjs
// @ts-check

import { $, nothrow } from 'zx'
import { Command } from 'commander'
import path from 'path';
import { NodeFire } from './nodefire/index.mjs'
import { which, CommandNotFoundError } from './util.mjs'

/**
 * create EPubSetting
 */
class EPubSetting {

    manuscript = 'book.md'
    title = 'title.txt'
    // name = 'book.epub'
    out = 'book.epub'
    css  = 'stylesheet.css'
    coverImage  = 'cover.jpg'
    target  = './'

    /**
     * deepcopy オブジェクト
     * @returns {EPubSetting}
     */
    clone() {
      const epubSetting = new EPubSetting();
      epubSetting.manuscript = this.manuscript;
      epubSetting.title = this.title;
      epubSetting.out = this.out;
      epubSetting.css = this.css;
      epubSetting.coverImage = this.coverImage;
      epubSetting.target = this.target;
      return epubSetting;
    }
    // out という名前のpathとtargetという名前のパス
}


/**
 * クラスYの説明
 */
class Ebook {

  /** @types {EPubSetting} */
  setting = new EPubSetting();

  /** 
   * @param {string | undefined} manuscript
   * @param {string | undefined} title
   * @param {string | undefined} out
   * @param {string | undefined} css
   * @param {string | undefined} coverImage
   * @param {string | undefined} target
   * @param {EPubSetting | undefined} setting
  */
  #setSetting(manuscript = undefined, title = undefined, out = undefined, css = undefined, coverImage=undefined, target=undefined, setting=undefined){

    if (setting === undefined) {
      setting = this.setting.clone();
    }

    // set variable from parameter.
    if (manuscript !== undefined) {
      setting.manuscript = manuscript;
    }
    if (title !== undefined) {
      setting.title = title;
    }
    if (out !== undefined) {
      setting.out = out;
    }
    if (css !== undefined) {
      setting.css = css;
    }
    if (coverImage !== undefined) {
      setting.coverImage = coverImage;
    }
    if (target !== undefined) {
      setting.target = target;
    }
    return setting
  }

  /**
   * 異常なものがないか確認。
   * @throws {CommandNotFoundError} command not exists.
   */
  check() {
    which('pandoc');
  }

  /** 
   * craete epub file from markdown.
   * @param {string | undefined} manuscript
   * @param {string | undefined} title
   * @param {string | undefined} out
   * @param {string | undefined} css
   * @param {string | undefined} coverImage
   * @param {string | undefined} target
   * @param {EPubSetting | undefined} setting
   * @throws {Error} command not exists.
  */
  async md2epub(manuscript = undefined, title = undefined, out = undefined, css = undefined, coverImage=undefined, target=undefined, setting=undefined){
    // 異常なものがないか確認。
    try {
      this.check();
    } catch (error) {
      /** @type {CommandNotFoundError} */
      const err = error;
      console.error(err.message);
      throw err;
    }
    this.setting = this.#setSetting(
        manuscript
        ,title
        ,out
        ,css
        ,coverImage
        ,target
        ,setting
    )

    const args = [
      '-f','markdown'
      ,'-t', 'epub3'
      ,path.join(this.setting.target, this.setting.manuscript)
      ,path.join(this.setting.target, this.setting.title)
      ,'-o', path.join(this.setting.target, this.setting.out)
      ,'--css', path.join(this.setting.target, this.setting.css)
      ,'--toc', '--toc-depth=2'
      ,`--epub-cover-image=${path.join(this.setting.target, this.setting.coverImage)}`
    ]
    try {
      await $`pandoc ${args}`;
    } catch (p) {
      console.log(`Exit code: ${p.exitCode}`)
      console.log(`Error: ${p.stderr}`)
    }
  }

  /**
   * execute command from terminal
   */
  invoke() {
    const program = new Command();
    program
      .option('-f, --function-name <YourValue>', 'you want to execute function')
      .option('-o, --out <YourValue>', 'Rectangle hight')
      .option('-t, --title <YourValue>', 'Rectangle hight')
      .option('-t, --target <YourValue>', 'Rectangle hight')
      .option('-c, --css <YourValue>', 'Rectangle width')
      .option('-c, --cover-image <YourValue>', 'Rectangle width');

    program.parse(process.argv);
    // key-valueで値がくるので、
    const options = program.opts();

    switch (options.functionName) {
      case 'md2epub':
        this.md2epub(
          options.manuscript
          ,options.title
          ,options.out
          ,options.css
          ,options.coverImage
          ,options.target
        );
        break;

      default:
        console.log("execute unkown functions.");
    }
  }
}

// main関数の代わり
if (process.env.__MAIN__ === __filename) {
  NodeFire(Ebook)
}

File: python3/src/ebook.py

from collections import namedtuple
import sys
import fire
from .version import Version
import subprocess
from copy import deepcopy
from subprocess import CalledProcessError
from dataclasses import dataclass
from os import path

@dataclass
class EPubSetting:
    """create EPubsetting
    """
    manuscript: str = 'book.md'
    title: str = 'title.txt'
    # name: str = 'book.epub'
    out: str = 'book.epub'
    css: str = 'stylesheet.css'
    cover_image: str = 'cover.jpg'
    target: str = './'
    # out という名前のpathとtargetという名前のパス

@dataclass
class Ebook:
    """This class is creating for epub."""

    setting = EPubSetting()

    def __setSetting(self, manuscript: str=None, title: str=None, out: str=None, css: str=None, cover_image: str=None,target: str =None, setting: EPubSetting=None) -> EPubSetting:

        if setting is None:
            setting = deepcopy(self.setting)

        # set variable from parameter.
        if manuscript:
            setting.manuscript = manuscript
        if title:
            setting.title = title
        if out:
            setting.out = out
        if css:
            setting.css = css
        if cover_image:
            setting.cover_image = cover_image
        if target:
            setting.target = target

        return setting

    def md2epub(self, manuscript: str=None, title: str=None, out: str=None, css: str=None, cover_image: str=None,target: str =None, setting: EPubSetting=None) -> None:
        """initialize Youtube Class.

        This class Youtube Operating Class

        Args:
            x: int variable.
            y: int valiable.

        Returns:
            None
        
        Raises:
            CalledProcessError: An error occurred executing shell script.
        """
        self.setting = self.__setSetting(
            manuscript
            ,title
            ,out
            ,css
            ,cover_image
            ,target
            ,setting
        )

        try:
            subprocess.run([
                'pandoc','-f','markdown'
                ,'-t', 'epub3'
                ,path.join(self.setting.target, self.setting.manuscript)
                ,path.join(self.setting.target, self.setting.title)
                ,'-o', path.join(self.setting.target, self.setting.out)
                ,'--css', path.join(self.setting.target, self.setting.css)
                ,'--toc', '--toc-depth=2'
                ,'--epub-cover-image={}'.format(path.join(self.setting.target, self.setting.cover_image))
            ],check=True
            ,capture_output=True)
        except CalledProcessError as err:
            print(err.stderr, file=sys.stderr)

if __name__ == '__main__':
    version = Version()
    if not version.check():
        print(
            'python version greater than {1}.{2}'.
                format(version.setting.major,version.setting.minor)
            ,file=sys.stderr
        )
        exit(1)

    fire.Fire(Ebook)
    # python aaa.py add 10 20
    # python multiply 10 20
